### DISCUSSION TITLE
Consider making it easier to add new TypeMapper instances based on old types

### DISCUSSION THREAD

**Discussion by felher** [com-lihaoyi/scalasql#25] [Ideas]
Hey all,

in my code bases I use a lot of new types, often just wrapping a more fundamental type like string or int. I assume that's rather common. If I want to use such a type, I think I currently have two ways to do so:

1. Create a `TypeMapper` from scratch, which requires at least a bit of understanding of JDBC/SQL. For example, for a `String`-based type you have to decide between `CHAR`, `VARCHAR` and `LONGVARCHAR`. Also, you have to know what a `cast[T]` is in SQL.
2. Piggyback on some `TypeMapper` already there, delegating to it and maybe casting the type somewhere.

Both of them involve a bit of boilerplate and don't really signal to the user "you're doing it right".

I suggest adding a simple method to `TypeMapper` or its companion object. Maybe something like:

```scala
object TypeMapper:
  def basedOn[T, U](tm: TypeMapper[T])(toBase: U => T, fromBase: T => U): TypeMapper
```

Apart from being less boilerplate, the user needs to know basically nothing about SQL/JDBC to know that what they have done is valid. Also, it makes it easier to not care about the dialect you are currently using, because a implicit for the base type should be in scope whenever `DialectTypeMappers` are in scope, so it should naturally use the correct one.

Any thoughts? Worth it, not worth it?

Felix

**Comment by lihaoyi** [comment]
Sounds reasonable. uPickle has this and calls it `bimap`

**Reply by lihaoyi** [reply]
Here's a PR implementing it https://github.com/com-lihaoyi/scalasql/pull/27

**Reply by lihaoyi** [reply]
Should be released in Scalasql 0.1.8

**Reply by felher** [reply]
Great! Thank you :pray:

**Reply by aboisvert** [reply]
Another thanks from the peanut gallery!  I will most likely use it as well.
